### ISSUE TITLE
Feature Request: We should be able to cycle between opened media in a chat - reported by @sig5

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#7862]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Problem:
It will be a nice UX touch for the user, adding convenience while finding an old media file.

## Solution: 
Add an option to cycle between media when opened via swipe/arrow keys/buttons.

## Action Performed:
1. Navigate to any chat 
2. Send multiple attachment
3. Try to navigate through older attachments/media 

## Expected Result:
User should be able to view older media in some preview 

## Actual Result:
User is unable easily to see older media 

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.39-1
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

**Expensify/Expensify Issue URL:**
**Issue reported by:** @sig5
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1644597095180509

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @trjExpensify (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @megankelso (`Design`), see [these Stack Overflow questions](https://stackoverflow.com/c/expensify/search?q=%5Bchore%5D%5Bteam-design%5D+is%3Aquestion&searchOn=1) for more details.

**Comment by trjExpensify** [comment]
Oh, this is cool! What are we thinking in terms of the UI for the controls for this? We should probably loop in design before moving this on to further, so I'm adding the `design` label for now and dropping the priority to `weekly`. After which, we can update the OP. 

In the meantime, here are some quick thoughts:

- When the member is on the most recent attachment in the chat, there's only a left arrow icon (centered and left-aligned?) in the preview screen. 
- If the member is previewing an attachment "further back" in the history, there will also be a right arrow icon (centered and right-aligned?) to navigate "forward" to the next attachment preview
- On desktop/Web: The arrow can be clicked to navigate to the next attachment preview, or the left/right arrow keys on the keyboard can be used
- On mobile/mWeb: The arrow can be tapped or the member can swipe left/right on the preview to navigate to the next attachment preview depending on the direction

- Should we give members the ability to dismiss/hide the navigation arrows in some capacity? On web/desktop I can see maybe only revealing the arrow(s) on hover? On smaller screens like mobile/mWeb they might get in the way, if we don't offer the ability to tap to dismiss/hide them in some capacity?

CC: @shawnborton

**Comment by trjExpensify** [comment]
Yeah, good question. I personally prefer being able to cycle through all attachments across all messages in a chat. The limit on message + threaded messages on Slack desktop is always a bit of a pain, IMO.

A neat thing WhatsApp does if you exit out of the attachment is navigate you back in the chat history to where it was originally posted for the added context in and around when it was sent. 

Posted on the original thread [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1646089399118769?thread_ts=1644597095.180509&cid=C01GTK53T8Q).

**Comment by trjExpensify** [comment]
Followed up in the slack thread, sorry I missed the replies! 😅

**Comment by trjExpensify** [comment]
Pulling a couple of things out of the [Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1644597095180509).. 

- Is it only the attachments on a given message/thread, or could you cycle through all attachments across all messages in a chat?

We don't have threads yet, and similarly one message is one attachment, so cycling through all attachments across all messages in a chat seems to be where we're at here?

- Should we include a horizontal scrollview like WhatsApp? 

I think we can probably descope this for now, though I agree it would be nice for speed scrolling. Especially if you know you're trying to get back to a point in time looking for something. 

- Should the carousel of attachments "wrap". i.e: if I have the last attachment open, and I press ArrowLeft am I taken to the first attachment or does nothing happen? Is the > in the carousel disabled, or does it take me back to the beginning?

As I mentioned in the thread, I don’t have a strong opinion on looping back to the most recent/oldest when you reach the end/beginning. I think we’d need to make that clear if we allowed for it though, which presumably means including some indication of date/time as well. Otherwise you'd have the perception of infinitely scrolling back, unaware that you've looped back to the start.

**Comment by trjExpensify** [comment]
Nice, works for me! 👍

**Comment by trjExpensify** [comment]
Okay cool, so what are the next steps here? We're ready to mock it up at this point, right?

**Comment by shawnborton** [comment]
That makes sense to me - looks like this is already assigned to @megankelso.

@trjExpensify can you update the original comment to reflect what the desired solution/expected result should be, and then Megan can make some mocks? I think we'll just need some left/right arrows to exist in the attachment viewer modal.

**Comment by trjExpensify** [comment]
>On mobile, we could just do it where a single tap on the image hides/shows the arrows?

Works for me! 👍

**Comment by megankelso** [comment]
some ideas for this 
<img width="678" alt="Screen Shot 2022-04-01 at 9 51 07 AM" src="https://user-images.githubusercontent.com/49413768/161307687-20922d36-f3ae-4240-b8c1-5ababb9b92de.png">

<img width="1041" alt="Screen Shot 2022-04-01 at 9 44 37 AM" src="https://user-images.githubusercontent.com/49413768/161307216-7d1dddde-3787-4c87-81d7-7688bc6e5392.png">

the hover/tap made me think y'all were suggesting the arrows as an overlay, would probably need to style differently but on first pass I prefer them not appearing over the image. curious for thoughts

<img width="336" alt="Screen Shot 2022-04-01 at 9 46 47 AM" src="https://user-images.githubusercontent.com/49413768/161307467-d7d3d66e-f118-4b10-9913-a6876c50f650.png">

**Comment by shawnborton** [comment]
I think it's pretty standard that the arrows would always be on the left & right edges of the screen, so I prefer those mockups in comparison to the mockups where the arrows are below the image. 

My concern with using white arrows is that if the attachment was white, you wouldn't see the arrow. Could we see a version where the arrows just use our standard button styling? Perhaps even borrowing some of the round styles you used for the video call mocks you made.

**Comment by shawnborton** [comment]
Also a small nitpick: can you update the screenshots to use the current attachment modal styling? It has a subheader for the filename as well as a bottom border:

<img width="1196" alt="image" src="https://user-images.githubusercontent.com/2319350/161323903-af743c71-82b0-4903-82cc-1af61ab3c6a8.png">

**Comment by megankelso** [comment]
I'm not sure what our standard button styling is but here are the ones I know of. I think the icon alone looks best but even the grey disappears on some images (big time on this one)

<img width="889" alt="Screen Shot 2022-04-01 at 12 26 24 PM" src="https://user-images.githubusercontent.com/49413768/161328967-6e0e1bee-76f1-48f2-9df4-416a28f900cf.png">

**Comment by megankelso** [comment]
also, do they the arrows in the circles feel a bit big to you? there's at 20 px but I wonder if we could bump down to 16 to allow for a smaller circle.

**Comment by shawnborton** [comment]
You can find all of our standard button styles in our component library, here is for reference though:
![image](https://user-images.githubusercontent.com/2319350/161329956-d4446f4e-cf8d-4bd0-86c3-c527b2ae174b.png)

So you have the sizes of:
- Small: 28px
- Medium: 40px
- Large: 52px

I think we'd want these to be at least medium size so they are reasonably tappable on mobile devices. So that being said, 40px button with a 20px icon might feel good. 

I think having a gray background is a good idea because this way we guarantee that the arrow will be visible on this gray background, even if the gray gets lost on the image. Also, this way we stay consistent and we don't introduce a new button style.

**Comment by megankelso** [comment]
here's the 40/20 button size. also it appears vertical photos may not appear with a border on mobile, so I updated that.

<img width="337" alt="Screen Shot 2022-04-01 at 12 46 16 PM" src="https://user-images.githubusercontent.com/49413768/161331520-7cf317ff-989b-4c7a-b348-b4dd593d8299.png">


I personally think these looks pretty chunky and prefer the arrows alone below the image. For mobile, I assume most people will swipe to see other attachments so I think including these are mostly to show you can scroll.

I can understand how the 40/20 buttons are more standard though. If we do the tap to hide/show, perhaps they won't feel so obtrusive.

**Comment by shawnborton** [comment]
Personally I think that looks great. Again, with arrows only, they can easily get lost depending on the image that is behind them so we very easily run into a usability issue. This also feels consistent with other button styles.

**Comment by megankelso** [comment]
I was suggesting the arrows go below the image not on top. 

Anyway, understand this solution feels more consistent. Thinking web would look something like this then
<img width="1043" alt="Screen Shot 2022-04-01 at 1 10 13 PM" src="https://user-images.githubusercontent.com/49413768/161334645-78e99c51-12a2-4d61-a80a-ed04c7ad8979.png">

**Comment by shawnborton** [comment]
For web, I was thinking the arrows would be closer to the edge of the screen. Especially since the attachment could be wider. Here is what Slack does for reference:
<img width="668" alt="image" src="https://user-images.githubusercontent.com/2319350/161335730-5b8f98f0-0657-49ab-88a3-3a1aded6eba8.png">


Thoughts?

**Comment by megankelso** [comment]
yeah we could move them over

<img width="1043" alt="Screen Shot 2022-04-01 at 1 23 46 PM" src="https://user-images.githubusercontent.com/49413768/161336194-4d8af415-7ad6-4068-b82a-7ba4cd92fc07.png">

**Comment by shawnborton** [comment]
Cool, I think that makes more sense. Thanks!

**Comment by megankelso** [comment]
ok for easy reference, mockups here! I think this is all we need.

<img width="375" alt="Mobile - Sidebar Open" src="https://user-images.githubusercontent.com/49413768/161339984-1abf990d-5a8a-4f3c-a1e2-8a1d74cad72e.png">
<img width="1208" alt="web attach" src="https://user-images.githubusercontent.com/49413768/161339988-09a17730-3c18-4d13-8bae-42cc342f3268.png">

**Comment by trjExpensify** [comment]
Purrrrfect, thanks! 😸 

Updated the OP to include the mocks and passing this over to `engineering` triage before going external. 👍

**Comment by alex-mechler** [comment]
I didnt realize how much I needed this before I saw this issue 😅 

Looks good to go external!

**Comment by trjExpensify** [comment]
Job on Upwork here: https://www.upwork.com/jobs/~011ca7e66b62951030

**Comment by Santhosh-Sellavel** [comment]
@trjExpensify Let's Double this one!

cc: @chiragsalian

**Comment by trjExpensify** [comment]
Yep, back from OoO today. Doubled [the job](https://www.upwork.com/jobs/~011ca7e66b62951030) on Upwork.

**Comment by chiragsalian** [comment]
Waiting on proposals.

**Comment by Santhosh-Sellavel** [comment]
I think it's time to double this!
cc: @trjExpensify @chiragsalian

**Comment by trjExpensify** [comment]
Doubled on Upwork [here](https://www.upwork.com/jobs/~011ca7e66b62951030).

**Comment by Santhosh-Sellavel** [comment]
Let's double this one up @trjExpensify.

cc: @chiragsalian

**Comment by trjExpensify** [comment]
Doubled!.

**Comment by JediWattson** [comment]
# Proposal

## Data flow

Use the navigator to retrieve the reportId from within the ImageRenderer. 
`Navigation.getActiveRoute().replace('/r/', "")`

This allows the modal to use Onyx and retrieve the reportActions for the proper id. From here a function called `show` is called when an image is pressed inside a chat, which can be used to create an array of attachment urls from the reportActions data. The array's findIndex is called to get the matching index of the current url and a page is set up to track the index. An event on press for the arrow will check if the page is within bounds and iterate along with changing the sourceURL on the state.

## UI

The arrows are set up using a View component with flexbox and absolute position, they sit next to the attachment view with a boolean that disables them is the reportId doesn't exist. I think there might a couple places I need to add these arrows so I might just create a function to reduce code. 

## The Story so Far

I know the arrows don't match the previous designs, but I wanted to get the positioning right before exploring what's available with the icons and any use cases I should cover with disabling them.

<img width="1014" alt="Screen Shot 2022-05-14 at 2 23 15 PM" src="https://user-images.githubusercontent.com/6119181/168443944-6733d34f-132b-4d0b-b698-e443c021902a.png">

**Comment by Santhosh-Sellavel** [comment]
@JediWattson 
Seems you are new here, welcome! 
Thanks for your interest to solve this, the results look promising!

Would you please elaborate more on technical code changes? so that I can review the proposal.

Look at this past [jobs](https://github.com/Expensify/App/issuesq=is%3Aissue+is%3Aclosed+label%3A%22Awaiting+Payment%22) as a reference to write proposals! 

And If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email [contributors@expensify.com](mailto:contributors@expensify.com) to request to join our Slack channel!

**Comment by Santhosh-Sellavel** [comment]
@JediWattson
I appreciate your effort, Sorry I am not able to follow you completely.
Can share the full proposal or elaborate a bit more clearly. 

Example:
1. Let's make a new Carousel Component to cycle to the images.
```js
  // Full Code for the component
  // Overview of component code snippets
```
2. Use it in the `Existing`  component Image, making the following changes.
```js
  // Changes snippet and elaborate why if needed.
```
.....

**Comment by JediWattson** [comment]
# Proposal

Sure thing, I think I've been looking at bug proposals when I should have picked out a feature from the link you posted. I certainly don't mind trying to get this right.

1. Make a Carousel component to cycle through images. (I can change the name to AttachmentCarousel)
```js
/// imports up here

function AttachmentsNav({ reportActions, onArrowPress, sourceURL }) {
    const actionsArr = Object.values(reportActions)
    const [page, setPage] = useState(-1)
    const [attachments, setAttachments] = useState([])
    useEffect(function(){
        const nextAttachments = actionsArr.reduce((arr, rep, index) => {
            const matches = CONST.REGEX.ATTACHMENT_SRC.exec(rep.originalMessage?.html)
            if(matches){
                if(matches[1].includes(sourceURL))
                    setPage(arr.length)        
                const url = matches[1].replace(
                    CONFIG.EXPENSIFY.EXPENSIFY_URL,
                    CONFIG.EXPENSIFY.URL_API_ROOT,
                )
                arr.push(url)
            }                                    
            return arr
        }, [])
        setAttachments(nextAttachments)

    }, [actionsArr.length])

    function handlePress(isBack){
        if(isBack ? page-1 < 0 : page+1 === attachments.length) return        
        const nextIndex = isBack ? page-1 : page+1        
        onArrowPress(attachments[nextIndex])
        setPage(nextIndex)        
    }

    return (
        <View style={{ width: "90%", position: "absolute",  justifyContent: "space-between", alignItems: "center", flexDirection: "row" }}>
            <Pressable onPress={() => handlePress(true)} style={{ cursor: "pointer" }}>
                <Icon src={BackArrow} height={42} width={42} />
            </Pressable>        
            <Pressable onPress={() => handlePress()} style={{ cursor: "pointer" }}>
                <Icon src={ArrowRight} height={42} width={42} />
            </Pressable>
        </View>                            
    )    
}

export default withOnyx({
    reportActions: {
        key: ({reportId}) => `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportId}`,
        canEvict: true
    },
})(AttachmentsNav)
```

2. Use the existing component AttachmentModal, making the following changes

in the show function here:
https://github.com/Expensify/App/blob/75d9014c75befb7655e78653c65f4d2e0c0f0cdf/src/components/AttachmentModal.js#L227

I added reportId to the state update which I get using the Navigation route
```js
show: () => {                        
    const reportId = Navigation.getActiveRoute().replace('/r/', "")
    this.setState({ reportId, isModalOpen: true});
}
```

the carousel component will sit next to the AttachmentView component here
https://github.com/Expensify/App/blob/75d9014c75befb7655e78653c65f4d2e0c0f0cdf/src/components/AttachmentModal.js#L183

this will render if reportId exists
```js
{this.state.reportId && ( 
    <AttachmentNav 
        reportId={this.state.reportId}                                    
        onArrowPress={this.onArrowPress}
        sourceURL={this.state.sourceURL}
    />
)}
```

lastly, there's a callback function added to AttachmentModal called onArrowPress, this will set the new url from the carousel component. 
```js
onArrowPress(sourceURL){
    this.setState({sourceURL})
}
```

**Comment by Santhosh-Sellavel** [comment]
@JediWattson can you share `ATTACHMENT_SRC` regex?

**Comment by JediWattson** [comment]
That would look like:

```  js
{
    ...,
    ATTACHMENT_SRC: /data-expensify-source="([^"]+)"/
}
```
I found it in [this forum](https://laracasts.com/discuss/channels/laravel/take-src-url-from-img-with-regex) and modified it to look at data-expensify-source since that seemed unique to attachments that I found in this boolean
https://github.com/Expensify/App/blob/75d9014c75befb7655e78653c65f4d2e0c0f0cdf/src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js#L30

**Comment by Santhosh-Sellavel** [comment]
I'm not sure what I am missing pressing next or back not working, can you share a video?

**Comment by Santhosh-Sellavel** [comment]
@JediWattson 
The proposal looks promising so far, I'm not sure what I'm missing cycling part doesn't work for me! If you can the working video or rewrite the proposal in case you missed something like `ATTACHMENT_SRC`. it will be helpful for us to make a decision here!

**Comment by JediWattson** [comment]
Hey that's great!

I made a couple changes because I realized the name of the file isn't being updated

I change the regex to use groups and now I'm matching the src url and name of the file here:
``` js
{
    ATTACHMENT_DATA: /(data-expensify-source|data-name)="([^"]+)"/g
}
```

I changed the matches function to the string function matchAll [here's the docs on how this works](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/matchAll), and I'm now saving the name in an object. 
```js
    useEffect(function(){
        const nextAttachments = actionsArr.reduce((arr, rep) => {
            const matchesIt = rep.originalMessage?.html.matchAll(CONST.REGEX.ATTACHMENT_DATA)
            if(matchesIt){
                const matches = [...matchesIt]
                if(matches.length === 2){
                    const [src, name] = matches
                    if(src[2].includes(sourceURL))
                        setPage(arr.length)
                    const url = src[2].replace(
                        CONFIG.EXPENSIFY.EXPENSIFY_URL,
                        CONFIG.EXPENSIFY.URL_API_ROOT,
                    )
                    arr.push({src: url, file: {name: name[2]}})
                }
            }                                                    
            return arr
        }, [])
        setAttachments(nextAttachments)
    }, [actionsArr.length])
```
This object is sent to a callback from the parent AttachmentModal and sets the state for the file. 
```js
    onArrowPress({src, file}){
        this.setState({sourceURL: src, file})
    }
```

A couple more things here too. I removed the ternary operator using props to parse the file name from this
https://github.com/Expensify/App/blob/75d9014c75befb7655e78653c65f4d2e0c0f0cdf/src/components/AttachmentModal.js#L104
to this: 
```js
const fullFileName = lodashGet(this.state, 'file.name', '').trim();
```

and to prevent error, I defaulted the file in state here:
https://github.com/Expensify/App/blob/75d9014c75befb7655e78653c65f4d2e0c0f0cdf/src/components/AttachmentModal.js#L75

to use the props or an empty string like this:
```js
{
    file: {name: props.originalFileName || ""}
}
```

Also here's the video (with the name changing)

https://user-images.githubusercontent.com/6119181/169702382-a8a96980-f6d9-4cde-bd6a-87f716c7595d.mov

**Comment by Santhosh-Sellavel** [comment]
@chiragsalian 
@JediWattson Proposal looks good to me!
🎀 👀 🎀 C+ reviewed

@JediWattson 
If you haven’t checked out already, please go through the following docs.
[CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md)
[README.md](https://github.com/Expensify/App/blob/main/README.md)
[PR_REVIEW_GUIDELINES.md](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
[STYLE.md](https://github.com/Expensify/App/blob/main/STYLE.md)
[STYLING.md](https://github.com/Expensify/App/blob/main/STYLING.md)

Can we could just go with Class component, instead of Pure Components with useEffects/useState. Because it's simple & readable and should do the Job! 

Check out the above code guidelines do necessary changes before you mark PR ready!

Any comments @chiragsalian?

Thanks!

**Comment by chiragsalian** [comment]
Yup i like the proposal, including santhosh's suggestions. I would also recommend more verbose variables instead of `arr` and `rep`.
Feel free to create the PR @JediWattson.

**Comment by trjExpensify** [comment]
👋 I've sent you the offer on Upwork, @JediWattson!

**Comment by chiragsalian** [comment]
Santhosh is back from OOO on June 15th. As requested above he's asked if we can hold review until he's back. I'm okay with holding for a week. Let me know if you guys feel differently.

**Comment by trjExpensify** [comment]
Latest from @JediWattson here on [the PR](https://github.com/Expensify/App/pull/9279#issuecomment-1166654597). 
>I just got most of the changes here up. I just need to add the functionality for the arrows to show and not show

**Comment by Santhosh-Sellavel** [comment]
Any update on the PR @JediWattson?

**Comment by JediWattson** [comment]
Hey I should have everything up except making it swipable. I have some changes for that, but I haven't worked out a transform in Animation.View

**Comment by chiragsalian** [comment]
Not overdue, PR is still in progress.

**Comment by chiragsalian** [comment]
Any update for us @JediWattson. The PR is pretty close to the finish line. 
Friendly reminder of our [contribution guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md),
> Daily updates on weekdays are highly recommended. If you know you won’t be able to provide updates for > 1 week, please comment on the PR or issue how long you plan to be out so that we may plan accordingly. We understand everyone needs a little vacation here and there. Any issue that doesn't receive an update for 1 full week may be considered abandoned and the original contract terminated.

**Comment by JediWattson** [comment]
yes @chiragsalian I believe I got all the requirements in the PR for this. I'm just trying to record some demos on android that are taking a bit for me. I'll try to finish them this week-end

**Comment by trjExpensify** [comment]
Where are we at here, guys? 

- Did we get anywhere on figuring out why the arrows can't be toggled on mWeb (and both mWeb & Native for PDFs)? If not, let's start a discussion in #expensify-open-source for more 👀 if we need them!

**Comment by JediWattson** [comment]
That sounds like a great idea to me

**Comment by JediWattson** [comment]
I really don’t have the time to put into this anymore

**Comment by Santhosh-Sellavel** [comment]
@JediWattson 
> I really don’t have the time to put into this anymore

I have roughly tested your PR, seems you addressed most of the requirements. Don't let your effort go in vain. Please let us know your concerns, if possible we can work it out. Thanks!

**Comment by JediWattson** [comment]
I can try to do something’s this weekend

**Comment by trjExpensify** [comment]
Nice! Looks like you [got to the bottom of it](https://github.com/Expensify/App/pull/9279#issuecomment-1214247308), @JediWattson. 🎉 

@Santhosh-Sellavel is asking for the OP of the PR to be updated to reflect the changes, and then this should be ready to go?

**Comment by JediWattson** [comment]
Hey @trjExpensify that's what I'm hearing. I need about a day or two before I can get that up, but I will be making time for it for sure!

**Comment by trjExpensify** [comment]
Awesome yeah, thanks!

**Comment by JediWattson** [comment]
I just update the videos in the PR and updated the branch with main, hopefully that does it!

**Comment by Santhosh-Sellavel** [comment]
@chiragsalian @trjExpensify 

Based on my review,  I believe this job is definitely time consuming, requesting a price increase to $4k. That would be reasonably fair.

Thanks!

**Comment by trjExpensify** [comment]
I think it would be good to re-evaluate once the job is complete at this point.

**Comment by Santhosh-Sellavel** [comment]
Fair call!

**Comment by trjExpensify** [comment]
PR is still in review, with active discussion.

**Comment by trjExpensify** [comment]
Marc & Chirag have offered some additional thoughts in [the PR](https://github.com/Expensify/App/pull/9279#issuecomment-1262997992) for going in the direction of a `Carousel` component.

**Comment by JediWattson** [comment]
@Santhosh-Sellavel Yes during the week I plan to try and finish this

**Comment by trjExpensify** [comment]
Thanks, @JediWattson. I'm incredibly excited to get this one over the line! ❤️

**Comment by chiragsalian** [comment]
Any update for us @JediWattson. The PR is pretty close to the finish line.
Reminder of our [contribution guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) once again,

> Daily updates on weekdays are highly recommended. If you know you won’t be able to provide updates for > 1 week, please comment on the PR or issue how long you plan to be out so that we may plan accordingly. We understand everyone needs a little vacation here and there. Any issue that doesn't receive an update for 1 full week may be considered abandoned and the original contract terminated.

More than a week has passed with no commits on the PR or any update here so it's hard to tell if you are actively working on the issue.

**Comment by chiragsalian** [comment]
Looks like there is movement on the PR. Thanks @JediWattson. Let us know if you've got any questions. Excited to see this through.

**Comment by mallenexpensify** [comment]
@trjExpensify I got assigned to the tracking issue for 'polish' and this is on there, I'm going to _co-assign_ with you as BugZero/Contributor Manager. (Since it's moving along I don't want to 👢 you since you know what's going on and I don't :)

**Comment by mallenexpensify** [comment]
@Beamanator we have most image-related issues on hold in the #10894 tracking issue. This one is in-progress, the PR is here - https://github.com/Expensify/react-native-onyx/pull/198
Is there any reason we **shouldn't** keep moving forward here?

**Comment by Santhosh-Sellavel** [comment]
@mallenexpensify That https://github.com/Expensify/react-native-onyx/pull/198 is for a different issue

**Comment by Santhosh-Sellavel** [comment]
Here is the PR for this issue https://github.com/Expensify/App/pull/9279

**Comment by mallenexpensify** [comment]
eeeps, sorry.  Thanks for the catch @Santhosh-Sellavel

**Comment by trjExpensify** [comment]
Yep, we should get this over the line. @Santhosh-Sellavel, can we get an update on where we are at and can we have the PR merged by the end of the week? :excited:

**Comment by Santhosh-Sellavel** [comment]
Yeah, the changes requested are done. Waiting for re-review.

Not sure it will be able to merge this week. But mostly will try to get it merged by next week. Code Review is only in progress,  still testing needs to be done if there are no surprises we will be close next week!

**Comment by trjExpensify** [comment]
Oh hm, will it take 2 weeks to review the PR? That seems like quite a lengthy time. Can @chiragsalian get going on his initial code review while you finish up on the tests or something?

**Comment by Santhosh-Sellavel** [comment]
Yeah, we could that helps speed up!

**Comment by trjExpensify** [comment]
Cool, @chiragsalian feel free to take a look at the PR. In the meantime, where are you at with the tests @Santhosh-Sellavel?

**Comment by trjExpensify** [comment]
Alrighttty.. it's Monday! What's it going to take to get this PR merged this week?

**Comment by Santhosh-Sellavel** [comment]
@JediWattson Can you provide an update here, whenever PR is ready for review if you need input from us gives a bump! Let's try and get this merged soon!

**Comment by JediWattson** [comment]
I will take a look later today and finish up some changes I had started
last week!

**Comment by Santhosh-Sellavel** [comment]
Starting last week daily updates are mandatory, please provide update daily!

@trjExpensify Let's make this daily!

**Comment by mallenexpensify** [comment]
@trjExpensify , not sure who was BZ on this, I unassigned and left to you as I'll be OOO for a month

**Comment by JediWattson** [comment]
Yesterday I got the requested changes up and I’m waiting to hear feedback on them

**Comment by trjExpensify** [comment]
Thanks, @JediWattson! Looks like we have a conflict if you want to take a look in the meantime.

**Comment by trjExpensify** [comment]
@JediWattson, looks like we still have a conflict to resolve. Can you address it today please, so @Santhosh-Sellavel can provide his final review? Thanks!

**Comment by trjExpensify** [comment]
@chiragsalian is in the midst of the review, Melv.

**Comment by JediWattson** [comment]
I'll be taking on some of the comments within the next couple days just to keep everyone updated here

**Comment by trjExpensify** [comment]
We had another round of reviews this week, but still need a few changes to be worked on. Also, there are conflicts to resolve again now. Can we complete those and merge this PR next week, please? It has been in review for 4 months at this point. Thanks!

**Comment by JediWattson** [comment]
I’ll try Take Care of all that in the next few days

**Comment by Santhosh-Sellavel** [comment]
@JediWattson bump!

**Comment by JediWattson** [comment]
Understood!! Apologies I’ve been a bit busy but I’m always trying to make
some time to finish this up. Hopefully this evening I’ll get some changes
up!!!

**Comment by trjExpensify** [comment]
Hey @JediWattson - we have concerns over how long this PR has been in review with no updates or being any further to completion over the course of several months. I believe Chirag has reminded you of the expectations to provide daily updates previously. We also encourage you to discuss any challenges with your PRs in our public slack channel #expensify-open-source, since it's a great place to crowdsource opinions and thoughts to keep them moving along when needed.

We appreciate you might be busy or not have the time for this anymore since it was started a long time ago, so maybe it’s best we part ways on it? In recent months, we’ve [updated our contributing guidelines to expect a PR within 12 business days](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#payment-for-contributions) ahead of terminating contracts. If you do think you’ll be able to complete the work, given that the guideline is a bit newer, I’m happy to keep it open for another week until Monday, Dec 12th. After which, we’ll close the PR and terminate the job. 

Let me know what you’d like to do, and thanks for your understanding!

**Comment by JediWattson** [comment]
I'm having trouble with login. I filled out a form to hopefully fix it, but it seems like this is getting really really close to the finish line

**Comment by trjExpensify** [comment]
Hey @JediWattson I know you've had some trouble here with the ongoing API issues we're experiencing, so that's been making it a little bit challenging. Did you get your latest one whitelisted and can you confirm when the videos will be updated? Thanks!

**Comment by JediWattson** [comment]
I confirm they will be. I'll have to fill it out again right now, but I'll keep trying!

**Comment by JediWattson** [comment]
@trjExpensify I can confirm they will be. I'll have to fill it out again right now, but I'll keep trying!

**Comment by trjExpensify** [comment]
HNY everyone! Can someone bring me up to speed on what's left?

**Comment by Santhosh-Sellavel** [comment]
@JediWattson Can you give us what's left here, are all requested changes and related bugs fixed? 

Currently, there are some conflicts to be resolved, other than that can you give us what's left here, are all requested changes and related bugs addressed? 

Please let us know, we can continue reviewing and get this done soon!

**Comment by JediWattson** [comment]
I was going to resolve conflicts later this evening, but other than that I
think I’m waiting on feedback from my changes to replacing the host name
for ngrok, hopefully that’s all good and we’re set though!

**Comment by trjExpensify** [comment]
Updates happening in [the PR](https://github.com/Expensify/App/pull/9279#issuecomment-1385796302). @JediWattson, can you merge main to resolve the conflicts ASAP please?

**Comment by Santhosh-Sellavel** [comment]
I've already requested for price increase here https://github.com/Expensify/App/issues/7862#issuecomment-1239828493

We decided to evaluate the price when the issue is [complete](https://github.com/Expensify/App/issues/7862#issuecomment-1244486406). I.e when PR is merged!

**Comment by MelvinBot** [comment]
@JediWattson, @chiragsalian, @trjExpensify, @Santhosh-Sellavel Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by MelvinBot** [comment]
@JediWattson, @chiragsalian, @trjExpensify, @Santhosh-Sellavel Eep! 4 days overdue now. Issues have feelings too...

**Comment by MelvinBot** [comment]
@JediWattson, @chiragsalian, @trjExpensify, @Santhosh-Sellavel Now this issue is 8 days overdue. Are you sure this should be a Daily? Feel free to change it!

**Comment by trjExpensify** [comment]
PR review is ongoing Melvin, chill.

**Comment by MelvinBot** [comment]
@JediWattson, @chiragsalian, @trjExpensify, @Santhosh-Sellavel Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
@JediWattson, @chiragsalian, @trjExpensify, @Santhosh-Sellavel Huh... This is 4 days overdue. Who can take care of this?

**Comment by MelvinBot** [comment]
@JediWattson, @chiragsalian, @trjExpensify, @Santhosh-Sellavel 8 days overdue is a lot. Should this be a Weekly issue? If so, feel free to change it!

**Comment by Santhosh-Sellavel** [comment]
Close to completion!

**Comment by mallenexpensify** [comment]
This appears to be live on production?!??!?
https://www.screencast.com/t/iuWw0UHe3

**Comment by MelvinBot** [comment]
@JediWattson, @chiragsalian, @trjExpensify, @Santhosh-Sellavel Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by trjExpensify** [comment]
Oh hm, the title didn't update automatically. Done that! 

@Santhosh-Sellavel @chiragsalian @luacmartins what are the next steps on the follow-up items? Seems like the most pressing item is figuring out how to "go back" further versus only as far as reportActions we have loaded, right? 

I think we'd go ahead and create a new issue for that.

**Comment by chiragsalian** [comment]
> what are the next steps on the follow-up items?

Create an issue for each, discuss within each if they are worth it at this time, and then mark them all as external.

> Seems like the most pressing item is figuring out how to "go back" further versus only as far as reportActions we have loaded, right?

Personally yeah I would like to see this one implemented. It will involve some backend changes.

**Comment by trjExpensify** [comment]
>Create an issue for each, discuss within each if they are worth it at this time, and then mark them all as external.

What are those? I seem to recall the debounce improvement and going back further being the only ones worth continuing. 

>Personally yeah I would like to see this one implemented. It will involve some backend changes.

Okay cool, so shall we have an `internal` issue spun up for that, who's going to take it on?

**Comment by MelvinBot** [comment]
@JediWattson, @chiragsalian, @trjExpensify, @Santhosh-Sellavel Huh... This is 4 days overdue. Who can take care of this?

**Comment by MelvinBot** [comment]
@JediWattson, @chiragsalian, @trjExpensify, @Santhosh-Sellavel 6 days overdue. This is scarier than being forced to listen to Vogon poetry!

**Comment by trjExpensify** [comment]
Alrighty, I've sent an offers for the increased values:

@JediWattson $4k for the fix 
@Santhosh-Sellavel $4k for C+

**Comment by sig5** [comment]
I think I have not been paid the bounty for the 🐛

**Comment by MelvinBot** [comment]
📣 @sig5! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by sig5** [comment]
Contributor details
Your Expensify account email: singhalsakar@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01bd078f22eaeb58dd

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by sig5** [comment]
https://github.com/Expensify/App/blob/b1a08ca2e41e5f43f4865c6518a2813a21b4b514/CONTRIBUTING.md It wasn't the case when the issue had started. Seems like there was a change in rules in the meanwhile. But I think old rules should be honoured because it was reported then.

**Comment by sig5** [comment]
Bump###

**Comment by sig5** [comment]
Need to get done with this, So BUMP @trjExpensify

**Comment by trjExpensify** [comment]
Hey @sig5, easy on the caps lock, I've been out of office until today. 🙂 

I checked the feature request freeze and you're right it did come into play after the issue creation. I've sent you a contract for the bug report on Upwork.

**Comment by sig5** [comment]
Ayy, thanks man. Didn't mean to be rude if felt that way.

**Comment by trjExpensify** [comment]
All good, that's settled now.
